bisync: bug fixes and new features including --create-empty-src-dirs and equality check

nielash opened the pull request:
<!--
Thank you very much for contributing code or documentation to rclone! Please
fill out the following questions to make it easier for us to review your
changes.

You do not need to check all the boxes below all at once, feel free to take
your time and add more commits. If you're done and ready for review, please
check the last box.
-->

#### What is the purpose of this change?

Several `bisync` bug fixes and new features including `--create-empty-src-dirs` and equality check.
Described in (excruciating) detail [here](https://forum.rclone.org/t/bisync-bugs-and-feature-requests/37636). And further articulated in the [proposed documentation changes](https://github.com/nielash/rclone/blob/256892d85c0f13934988bd64731e19e119836be5/docs/content/bisync.md).

In summary:
* Fixed an [issue](https://forum.rclone.org/t/bisync-bugs-and-feature-requests/37636#:~:text=1.%20Dry%20runs%20are%20not%20completely%20dry) causing dry runs to inadvertently commit filter changes
* Fixed an [issue](https://forum.rclone.org/t/bisync-bugs-and-feature-requests/37636#:~:text=2.%20%2D%2Dresync%20deletes%20data%2C%20contrary%20to%20docs) causing `--resync` to erroneously delete empty folders and duplicate files unique to Path2
* `--check-access` is now enforced during `--resync`, preventing data loss in [certain user error scenarios](https://forum.rclone.org/t/bisync-bugs-and-feature-requests/37636#:~:text=%2D%2Dcheck%2Daccess%20doesn%27t%20always%20fail%20when%20it%20should)
* Fixed an [issue](https://forum.rclone.org/t/bisync-bugs-and-feature-requests/37636#:~:text=5.%20Bisync%20reads%20files%20in%20excluded%20directories%20during%20delete%20operations) causing bisync to consider more files than necessary due to overbroad filters during delete operations
* [Improved detection of false positive change conflicts](https://forum.rclone.org/t/bisync-bugs-and-feature-requests/37636#:~:text=1.%20Identical%20files%20should%20be%20left%20alone%2C%20even%20if%20new/newer/changed%20on%20both%20sides) (identical files are now left alone instead of renamed)
* Added [support for `--create-empty-src-dirs`](https://forum.rclone.org/t/bisync-bugs-and-feature-requests/37636#:~:text=3.%20Bisync%20should%20create/delete%20empty%20directories%20as%20sync%20does%2C%20when%20%2D%2Dcreate%2Dempty%2Dsrc%2Ddirs%20is%20passed)
* Added experimental `--resilient` mode to allow [recovery from self-correctable errors](https://forum.rclone.org/t/bisync-bugs-and-feature-requests/37636#:~:text=2.%20Bisync%20should%20be%20more%20resilient%20to%20self%2Dcorrectable%20errors)
* Added [new `--ignore-listing-checksum` flag](https://forum.rclone.org/t/bisync-bugs-and-feature-requests/37636#:~:text=6.%20%2D%2Dignore%2Dchecksum%20should%20be%20split%20into%20two%20flags%20for%20separate%20purposes) to distinguish from `--ignore-checksum`
* [Performance improvements](https://forum.rclone.org/t/bisync-bugs-and-feature-requests/37636#:~:text=6.%20Deletes%20take%20several%20times%20longer%20than%20copies) for large remotes
* Documentation and testing improvements

#### Was the change discussed in an issue or in the forum before?

Yes, mainly in this forum post:
https://forum.rclone.org/t/bisync-bugs-and-feature-requests/37636

It also fixes #6109 and addresses several other open issues: #6841 #5683 #5681 #5676 #5675 #5674

#### Checklist

- [x] I have read the [contribution guidelines](https://github.com/rclone/rclone/blob/master/CONTRIBUTING.md#submitting-a-new-feature-or-bug-fix).
- [x] I have added tests for all changes in this PR if appropriate.
- [x] I have added documentation for the changes if appropriate.
- [x] All commit messages are in [house style](https://github.com/rclone/rclone/blob/master/CONTRIBUTING.md#commit-messages).
- [x] I'm done, this Pull Request is ready for review :-)

ncw commented:
Would it be possible to break this down into a commit per fix/feature?

If the commit messages can link to either the forum or an rclone issue as well that would be perfect.

Reviewing a lot of changes all at once is very difficult :-(

I don't mind receiving this all as one pull request with lots of commits, but I'd suggest you start with a PR with one or a few of the simple fixes so we can get into the swing of things.

nielash commented:
Sure thing, sorry about that 😃

Btw, it looks like a lot more than it is, because the vast majority is docs and test file changes.

Also the tests all pass locally...looks like they are failing on GH because the test.log files got gitignored out of the golden directory. Will fix and resubmit. 😃

nielash commented:
Ok I broke this down into a commit per fix/feature and confirmed the [GitHub tests are passing now](https://github.com/nielash/rclone/actions/runs/4788341304).

Note that I left all the test changes together in one commit: https://github.com/rclone/rclone/commit/e1e5593bedd950a8ca750128c1d8d88924877a1c I don't think it's really possible to separate them due to the way the Bisync test suite is designed (it diffs an actual log with a golden log, so [just about any unrelated change you make will cause false positives](https://rclone.org/bisync/#updating-golden-results). I could make them each pass separately, but then they might not all pass together.)

If you want to start with just a few, I think these are the most important:

- https://github.com/rclone/rclone/pull/6971/commits/748d14bb9e9c378e0e01f35a00ed123ad0b8fb0a
- https://github.com/rclone/rclone/pull/6971/commits/eb3ee5593355171cccca91314b5eecc2a81788e4
- https://github.com/rclone/rclone/pull/6971/commits/d41cc7bae211784ebf7c1001cbaa7404eff18f97

But these are the simplest:

- https://github.com/rclone/rclone/pull/6971/commits/718d9d63ceea35573beab8a81932989250992329
- https://github.com/rclone/rclone/pull/6971/commits/bd1f31d4c70f51d25245da9beee50b381e19b771

ncw commented on cmd/bisync/deltas.go:
```diff
@@ -290,6 +316,24 @@ func (b *bisyncRun) applyDeltas(ctx context.Context, ds1, ds2 *deltaSet) (change
+		//propagate deletions of empty dirs from path2 to path1 (if --create-empty-src-dirs)
+		if b.opt.CreateEmptySrcDirs && !b.opt.Resync {
+			for _, s := range delete1.ToList() {
+				_, objerr := b.fs1.NewObject(ctx, s)
```
same comments as above!

ncw commented on cmd/bisync/deltas.go:
```diff
@@ -299,6 +343,24 @@ func (b *bisyncRun) applyDeltas(ctx context.Context, ds1, ds2 *deltaSet) (change
+		//propagate deletions of empty dirs from path1 to path2 (if --create-empty-src-dirs)
+		if b.opt.CreateEmptySrcDirs && !b.opt.Resync {
+			for _, s := range delete2.ToList() {
+				_, objerr := b.fs2.NewObject(ctx, s)
```
Duplicated code?

nielash commented:
Thanks so much for the excellent feedback! I'm happy to hear that you're considering this for the next release.

The `NewObject` thing is something I wondered about too, and wanted to ask your advice about. What would you recommend instead? Also, is there a simpler way of checking whether it's a file or a dir, without relying on `ErrorIsDir`? (As an aside, I'm presuming that bucket-based backends will not be using `--create-empty-src-dirs` in the first place, as the feature would be meaningless... but it does concern me that `NewObject` only returns `ErrorIsDir` "if possible without doing any extra work"...)

Regarding the tests -- I can do what you're requesting and I don't mind the extra work, but I do fear that the result you're looking for is not truly possible due to the unique way the bisync test suite is designed. I'll explain what I mean below, and if you're ok with this caveat I am happy to do it. (And apologies in advance if this is all stuff you already know -- just making sure we're on the same page.)

The bisync tests compare not only the resulting files but **also the debug log** to the respective golden versions. (This is why I had to remove `*.log` from `.gitignore`, as you noted.) [As documented here](https://rclone.org/bisync/#updating-golden-results), this means that just about anything you change will cause false positives, whether or not it's related to what's being tested. For this reason, the recommended workflow is to goldenize, check the difference carefully, and then commit if you're convinced that all the diffs are due either to intentional changes or false positives. (In my opinion, this is kind of a design flaw -- it has the upside of being an extremely sensitive test, but the downside of relying on humans to do the checking, and thus being more prone to human error. It is probably a little too easy, currently, for a future developer to accidentally turn a true positive into a false negative among all the noise.)

As an example, my "equality check" commit adds a line that prints `"Files are equal! Skipping."` to the log in certain scenarios. Let's assume for simplicity's sake that this is all it does. Every subsequent commit will then expect that line to be there, and fail if it is not. So, let's say that my next commit deals with something totally unrelated, and adds a new test for whatever it is. If you were to then revert the first commit, it would cause tests to fail for the second commit -- even though both commits passed all tests individually, and even though nothing is actually "broken". Removing `"Files are equal! Skipping."` from all the 'golden' logs might be enough to fix this, but git isn't going to know that, at least for the new test log that didn't exist prior to the second commit. (As another example: even merging the same commits in a different order would likely cause failures.)

So, I can give you a sequence of commits that each pass all their tests, but they will be dependent on their sequence, and therefore not necessarily cherry-pickable. If you were to later revert a commit from the middle of the pack, as you describe, **you'd likely have to also revert every commit that came after it** (or alternatively push a new commit to re-goldenize.) Does that make sense, and are you ok with that?

Assuming yes, what's your preferred way of receiving these revisions? Should I revert these commits and add new ones, or is it cleanest to just create a new branch?

All of your other comments made total sense and shouldn't be a problem.

Thanks!

nielash commented:
> Wonder if these extended changes might have an effect on #6572 too...

(I replied in the #6572 thread... TL;DR: I think it's fixed in the latest beta, and not specific to bisync.)

Also, @ncw just tagging you here in case you didn't see my previous post with some questions for you (no rush at all.)

callegar commented:
With the beta I am seeing this all the time:
```
ERROR : 01_EXAMINED/XXX/commenti_progetto.md.rclonelink: Failed to read link size: readlink /home/user/OneDrive/here/01_EXAMINED/XXX/commenti_progetto.md: invalid argument
```

ncw commented:
Sorry have been rather busy!

Will answer questions inline.

> Thanks so much for the excellent feedback! I'm happy to hear that you're considering this for the next release.

I think we might have to push until next release - sorry!

> The `NewObject` thing is something I wondered about too, and wanted to ask your advice about. What would you recommend instead? Also, is there a simpler way of checking whether it's a file or a dir, without relying on `ErrorIsDir`? (As an aside, I'm presuming that bucket-based backends will not be using `--create-empty-src-dirs` in the first place, as the feature would be meaningless... but it does concern me that `NewObject` only returns `ErrorIsDir` "if possible without doing any extra work"...)

Ideally you'd keep a the objects from the directory listing so you don't have to re-instantate them. This may not be possible of course, but in general `NewObject` should be avoided in favour of saving objects from listings. For most backends 1 `NewObject` takes about the same amount of time as 1 directory listing, and on some it takes longer (eg box).

> Regarding the tests -- I can do what you're requesting and I don't mind the extra work, but I do fear that the result you're looking for is not truly possible due to the unique way the bisync test suite is designed. I'll explain what I mean below, and if you're ok with this caveat I am happy to do it. (And apologies in advance if this is all stuff you already know -- just making sure we're on the same page.)
> 
> The bisync tests compare not only the resulting files but **also the debug log** to the respective golden versions. (This is why I had to remove `*.log` from `.gitignore`, as you noted.) [As documented here](https://rclone.org/bisync/#updating-golden-results), this means that just about anything you change will cause false positives, whether or not it's related to what's being tested. For this reason, the recommended workflow is to goldenize, check the difference carefully, and then commit if you're convinced that all the diffs are due either to intentional changes or false positives. (In my opinion, this is kind of a design flaw -- it has the upside of being an extremely sensitive test, but the downside of relying on humans to do the checking, and thus being more prone to human error. It is probably a little too easy, currently, for a future developer to accidentally turn a true positive into a false negative among all the noise.)

These are really what I'd call regression tests then...

> As an example, my "equality check" commit adds a line that prints `"Files are equal! Skipping."` to the log in certain scenarios. Let's assume for simplicity's sake that this is all it does. Every subsequent commit will then expect that line to be there, and fail if it is not. So, let's say that my next commit deals with something totally unrelated, and adds a new test for whatever it is. If you were to then revert the first commit, it would cause tests to fail for the second commit -- even though both commits passed all tests individually, and even though nothing is actually "broken". Removing `"Files are equal! Skipping."` from all the 'golden' logs might be enough to fix this, but git isn't going to know that, at least for the new test log that didn't exist prior to the second commit. (As another example: even merging the same commits in a different order would likely cause failures.)
> 
> So, I can give you a sequence of commits that each pass all their tests, but they will be dependent on their sequence, and therefore not necessarily cherry-pickable. If you were to later revert a commit from the middle of the pack, as you describe, **you'd likely have to also revert every commit that came after it** (or alternatively push a new commit to re-goldenize.) Does that make sense, and are you ok with that?

Yes that makes sense. I think the intellectual exercise of checking the diffs match what we expect after each change is probably worth doing. We can hopefully spot any problems with that commit.

Noted that it effectively makes the commits un cherry-pickable.

> Assuming yes, what's your preferred way of receiving these revisions? Should I revert these commits and add new ones, or is it cleanest to just create a new branch?

I'm happy for you to force push this branch with new commits. That is what I'd normally do. But it doesn't matter much to me so if you want to create a new branch which probably means a new PR that is fine. What I'd probably do is clone the branch as `bisync-fixes-and-features-old` to make a backup. Then I'd use revert all the code and use `git add -p` to add the parts of each bit, `git stash` then make the golden tests, check the diff then add to the commit. Then use git stash pop to get the remaining changes back and repeat for each feature. Then when done, force push here.

> All of your other comments made total sense and shouldn't be a problem.

:-)

nielash commented:
Very helpful, thank you! I think I've found a good solution that eliminates the need to use `NewObject`. It will take me some time to do all the testing, but I'll get it to you as soon as I can. Thanks!

nielash commented on docs/content/bisync.md:
```diff
@@ -398,38 +498,97 @@ This will be solved in a future release, there is no workaround at the moment.
 ### Renamed directories
 
-Renaming a folder on the Path1 side results is deleting all files on
+Renaming a folder on the Path1 side results in deleting all files on
```
Worth noting: #5690

nielash commented:
@ncw Ok, I think this is ready for you to take a look, whenever you have a chance. I think it addresses all of your comments. Each commit should now include all relevant documentation and testing changes, and each commit should compile and pass all tests.

The equality check and `--create-empty-src-dirs` commits are significantly revised to be much more efficient and avoid having to use `NewObject`. Which of the objects are dirs is now noted during the initial directory listing so we don't have to re-list for each object later. The equality check is now performed once, outside the loop, using the same underlying function as `check` (so it should be just as efficient.) We do the check only if we need to, and filter it for only the few files we're interested in.

Since this branch has gotten quite noisy with all the reversions, I also made a [new "clean" branch](https://github.com/nielash/rclone/tree/bisync-fixes-and-features-CLEAN) with just the "real" commits cherry-picked. Feel free to use whichever you prefer.

Thanks!

nielash commented:
(I [added a fix](https://github.com/rclone/rclone/pull/6971/commits/c00356b2f81dc4c17946ebb51a10ef1fce308cd7) for a small bug I found after a few weeks of real-world use.)

ncw commented:
Thanks @nielash for your hard work :-)

Can you rebase this on master and fix the conflict, please. (Its probably the fact I added a new parameter to the flags, you can just use `""` for the flag group).

If you can force push your clean branch over this one (with your new fix integrated into the correct commit + the rebase) - that would be my preferred way of working.

I'd like to merge this for v1.64 :crossed_fingers:

nielash commented:
I think it should be all set now 👍

ncw reviewed (approved):
Thank you @nielash that is a very nice set of patches :-)

I will merge them now.

Any chance you'd like to be the maintainer of rclone's `bisync` code from now on? Drop me an email to nick@craig-wood.com if you are interested :-)

nielash commented:
> Thank you @nielash that is a very nice set of patches :-)

Thanks! They've certainly improved my workflow and I hope others will find them helpful as well.

> Any chance you'd like to be the maintainer of rclone's bisync code from now on? Drop me an email to [nick@craig-wood.com](mailto:nick@craig-wood.com) if you are interested :-)

I'm definitely interested, depending on the time commitment you're looking for. I'll drop you an email 😃
